### ISSUE TITLE
Move from post meta to custom taxonomies for restriction data

### ISSUE THREAD

**Issue by tnorthcutt** [restrictcontentpro/restrict-content-pro#1736]
_This is an offshoot of https://github.com/restrictcontentpro/restrict-content-pro/issues/135_

@mindctrl and I discussed this a bit today. Right now, `rcp_get_restricted_post_ids()` does a [pretty big post meta query](https://github.com/restrictcontentpro/restrict-content-pro/blob/a00264b2001c9553f3138b1dc478b4913ff327d2/includes/misc-functions.php#L1099-L1140), which on large sites can be a problem.

Some very rudimentary testing (calling `rcp_get_restricted_post_ids()` in my theme's `page.php` template) with Query Monitor confirmed this query gets very slow pretty quickly, relative to number of posts in the database. At a few thousand posts, this got up to .25 seconds.

I'll run some additional tests to try and get a rough comparison to a taxonomy query that would be roughly what we'd use, and update this issue.

**Comment by tnorthcutt** [comment]
Update: some rudimentary testing suggests that with ~6k posts, using taxonomies instead of post meta results in about 1/5 the query time. Very rough, but a big enough difference to make further work on this worth exploring.

**Comment by tnorthcutt** [comment]
On hold for now per https://github.com/restrictcontentpro/restrict-content-pro/issues/1670#issuecomment-375914648

**Comment by tnorthcutt** [comment]
Revisiting this today, it seems my earlier conclusions about query performance were premature. I think I may have caused the initial testing to be slow by performing those queries immediately after using wp-cli to insert several thousand posts with associated post meta. My theory is that temporarily degraded my CPU performance which impacted the subsequent queries.

So far I haven't been able to reproduce the slow (~.25 second) queries using the (existing) postmeta method; today they're taking roughly the same amount of time as the taxonomy query method (~.05 seconds).

I'm still open to making this change in the interest of a cleaner data model, and perhaps those query times would be more impacted on a larger site with a five or six digit number of posts.

@mindctrl I'd love your input on this 🙂

**Comment by mindctrl** [comment]
@tnorthcutt we really only see it on crappy hosts that limit cpu or throttle mysql in some way (I've seen Hostgator throttle MySQL so that it rejects connections and causes data loss on other non-RCP sites), and/or sites with the five/six digit post counts. You should see a performance hit after 10k posts or some random high number like that.

**Comment by mindctrl** [comment]
I think the taxonomy query will be noticeably faster at scale and we could probably ditch `rcp_get_restricted_post_ids()` entirely.

I'll give this some more thought today/tomorrow and circle back around.

**Comment by tnorthcutt** [comment]
@mindctrl makes sense re: crappy hosts. Would line up perfectly with what I saw with degraded CPU performance.

How do you mean we could ditch that function? Do you mean in `rcp_hide_premium_posts()` we could perhaps directly set the taxonomy query params, rather than using `post__not_in` and passing an array of post IDs?

**Comment by mindctrl** [comment]
@tnorthcutt yep, that was my thinking. Haven't yet explored all of what that might mean.

**Comment by tnorthcutt** [comment]
@mindctrl I like that idea a lot. One less query to perform, plus ultimately using a taxonomy query _instead of_ a `post__not_in` parameter. I assume the `post__not_in` parameter is rather expensive, especially for a large array of post IDs.

I'll poke at it a bit soonish.

**Comment by tnorthcutt** [comment]
@mindctrl I've got a (very) WIP PR here: https://github.com/restrictcontentpro/restrict-content-pro/pull/1786

So far I believe it's working as expected; I'm going to do some more cleanup and testing today.

**Comment by tnorthcutt** [comment]
Scratch that; not working as desired yet 😬

**Comment by tnorthcutt** [comment]
Ok, ran into a complication (and figured out what it was): turns out postmeta values can be `0`, but a taxonomy term cannot have a slug of `0`.

This is because `wp_set_post_terms()` checks if `$tags` is `empty()`, and if so sets it to an empty array before calling `wp_set_object_terms()`.

This is a problem when we're trying to restrict a post to an access level as here: https://github.com/restrictcontentpro/restrict-content-pro/pull/1786/files#diff-0d14859708edbbbb633e1f7f51323123R186

@mindctrl I'm not sure what the workaround for this is... do we perhaps check for an access level of `0` and convert that to `zero`? That feels super hacky, but... I don't have another idea right now.

**Comment by mindctrl** [comment]
Hmm I think if someone sets it to `0 and higher`, we should just don't set the term for that. For back compat we could update `rcp_user_has_access()` to return a `0` in the case of it not being set.

**Comment by mindctrl** [comment]
Sorry, I didn't mean `rcp_user_has_access()`. I just meant that when we check for the term on the post, if it's not set we assume a `0` for comparison.

**Comment by tnorthcutt** [comment]
@mindctrl I just realized the existing code (which checks that a postmeta key of `rcp_access_level` exists and is not `None`) is basically pointless anyway, since `None` can never get saved as the meta value; `0` is what gets saved.

But to your point: I don't think that works (at least not if we want to keep parity with the current behavior). Here's why:

What we're trying to do here is check for restricted posts. If we treat a post with no `rcp_access_level` taxonomy term set the same as one with `0` set, we'll be getting a false positive – any post with no restrictions will return `true` and be considered restricted.

With the current codebase, a post that is set to require an access level of `0 and higher` will be considered restricted:

![restrictions](https://user-images.githubusercontent.com/796639/39071382-728c9464-44ac-11e8-8d83-54bd2db8ac50.png)

So, we can't simply say "don't consider `0  and higher` to be a restricted post" _if we want to maintain parity_.

But we also can't say "if it's not set, assume `0 and higher`" because then every unrestricted post would be assumed to be set to "require an access level of `0 and higher`".

This is hard to describe... am I making any sense?

**Comment by tnorthcutt** [comment]
Ok, I made some progress here, once I figure out the proper way to pass data to `wp_set_post_terms()`.

**Comment by mindctrl** [comment]
On the comment about Access Level and `0 and higher`. I don't think this option should even exist in the dropdown.

If you create a post and set the Access Level to `0 and higher`, we save it as to `rcp_access_level` as `0`, but that doesn't really make sense. There's no way for a user to get access level `0` unless someone does it programmatically.

We don't give you a way to set a user access level to `0`. The dropdown on the subscription level add/edit screen starts at `1`.

And the member edit UI doesn't even give you a way to set access level at all.

And several places in the code we check for a `0` so we can basically ignore it. See `RCP_Member::can_access()` and `rcp_user_has_access()`.

In the end, `0 and higher` is effectively the same thing as `1 and higher`.
